### PR TITLE
Add basic airspeed velocity (ASV) workflow

### PR DESCRIPTION
**Pull request by kathivashkiv**
There are two parts in this PR:
1) ASV benchmark suite (`benchmarks` dir + `asv.conf,json` config file)
2) GitHub Actions workflow, which pushes ASV run result to the `sunpy/sunpy-benchmarks` repository. Follow this link to check this results: https://github.com/sunpy/sunpy-benchmarks/tree/main/results

### CODE REVIEW

**Comment by kathivashkiv** [comment]
@Cadair @nabobalis Please, have a look

File: .github/workflows/asv-initial.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+name: asv-initial-workflow
+
+on:
+  push:
+    branches: 
+      - asv-ci
+      - master
+jobs:
+  asv-run:
+    runs-on: ubuntu-18.04
```
Comment (Cadair) [inline .github/workflows/asv-initial.yml:11]: Should we bump this to 20.04?

File: .github/workflows/asv-initial.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
+# This is an initial workflow to run airspeed velocity
+name: asv-initial-workflow
```
Comment (Cadair) [inline .github/workflows/asv-initial.yml:2]:
```suggestion
name: asv-benchmarks
```

File: benchmarks/benchmarks_for_map.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+
+class TimeSuite:
+    """
+    An example benchmark that times the performance of sunpy map
+    creation
+    """
+    def setup(self):
+        self._map = sunpy.data.sample.AIA_171_IMAGE
+
```
Comment (Cadair) [inline benchmarks/benchmarks_for_map.py:15]: Is there any reason to do this in a setup method? (other than just to show how it's done)? there isn't any compute being done here as the sample data would be downloaded on import of this file?

File: asv.conf.json [hunk 4]
```diff
@@ -0,0 +1,164 @@
+{
+    // The version of the config file format.  Do not change, unless
+    // you know what you are doing.
+    "version": 1,
+
+    // The name of the project being benchmarked
+    "project": "asv_sunpy_project",
```
Comment (Cadair) [inline asv.conf.json:7]: should this not just be:
```suggestion
    "project": "sunpy",
```

File: asv.conf.json [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+    // The project's homepage
+    "project_url": "http://sunpy.org/",
+
+    // The URL or local path of the source code repository for the
+    // project being benchmarked
+    "repo": "https://github.com/sunpy/sunpy",
+    //"repo": "..",
```
Comment (Cadair) [inline asv.conf.json:15]:
```suggestion
```

File: asv.conf.json [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+    // Customizable commands for building, installing, and
+    // uninstalling the project. See asv.conf.json documentation.
+    //
+    // "install_command": ["in-dir={env_dir} python -mpip install {wheel_file}"],
+    // "uninstall_command": ["return-code=any python -mpip uninstall -y {project}"],
+    // "build_command": [
+    //     "python setup.py build",
+    //     "PIP_NO_BUILD_ISOLATION=false python -mpip wheel --no-deps --no-index -w {build_cache_dir} {build_dir}"
+    // ],
```
Comment (Cadair) [inline asv.conf.json:30]: Shall we remove sections of the config we aren't using to make it easier to read?
```suggestion
```

File: asv.conf.json [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+    // List of branches to benchmark. If not provided, defaults to "master"
+    // (for git) or "default" (for mercurial).
+    "branches": ["master"], // for git
+    // "branches": ["default"],    // for mercurial
```
Comment (Cadair) [inline asv.conf.json:35]:
```suggestion
```

File: asv.conf.json [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    // determined by looking for tools on the PATH environment
+    // variable.
+    "environment_type": "virtualenv",
+
+    // timeout in seconds for installing any dependencies in environment
+    // defaults to 10 min
+    //"install_timeout": 600,
+
+    // the base URL to show a commit for the project.
+    // "show_commit_url": "http://github.com/owner/project/commit/",
```
Comment (Cadair) [inline asv.conf.json:55]:
```suggestion
    "show_commit_url": "https://github.com/sunpy/sunpy/commit/",
```

File: asv.conf.json [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    // pip (with all the conda available packages installed first,
+    // followed by the pip installed packages).
+    //
+    "matrix": {
+    	 "extension-helpers": [],
+     	 "matplotlib": [],
+         "numpy": [],
+	 "scipy": [],
+         "six": [],        // test with and without six installed
+    //     "pip+emcee": [""],   // emcee is only available for install with pip.
```
Comment (Cadair) [inline asv.conf.json:81]:
```suggestion
```

File: asv.conf.json [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+    // the package name by 'pip+', and the package will be installed via
+    // pip (with all the conda available packages installed first,
+    // followed by the pip installed packages).
+    //
+    "matrix": {
+    	 "extension-helpers": [],
+     	 "matplotlib": [],
+         "numpy": [],
+	 "scipy": [],
+         "six": [],        // test with and without six installed
```
Comment (Cadair) [inline asv.conf.json:80]:
```suggestion
```

File: asv.conf.json [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    // installed. If the package to be tested is only available from
+    // PyPi, and the 'environment_type' is conda, then you can preface
+    // the package name by 'pip+', and the package will be installed via
+    // pip (with all the conda available packages installed first,
+    // followed by the pip installed packages).
+    //
+    "matrix": {
+    	 "extension-helpers": [],
+     	 "matplotlib": [],
+         "numpy": [],
```
Comment (Cadair) [inline asv.conf.json:78]:
```suggestion
       "numpy": [],
```

File: asv.conf.json [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+    // The thresholds for relative change in results, after which `asv
+    // publish` starts reporting regressions. Dictionary of the same
+    // form as in ``regressions_first_commits``, with values
+    // indicating the thresholds.  If multiple entries match, the
+    // maximum is taken. If no entry matches, the default is 5%.
+    //
+    // "regressions_thresholds": {
+    //    "some_benchmark": 0.01,     // Threshold of 1%
+    //    "another_benchmark": 0.5,   // Threshold of 50%
+    // },
```
Comment (Cadair) [inline asv.conf.json:163]:
```suggestion
```

File: asv.conf.json [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+    //     {"python": "3.2", "sys_platform": "win32"}, // skip py3.2 on windows
+    //     {"environment_type": "conda", "six": null}, // don't run without six on conda
+    // ],
+    //
+    // "include": [
+    //     // additional env for python2.7
+    //     {"python": "2.7", "numpy": "1.8"},
+    //     // additional env if run on windows+conda
+    //     {"platform": "win32", "environment_type": "conda", "python": "2.7", "libpython": ""},
+    // ],
```
Comment (Cadair) [inline asv.conf.json:115]:
```suggestion
```

File: asv.conf.json [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+    // the base URL to show a commit for the project.
+    // "show_commit_url": "http://github.com/owner/project/commit/",
+
+    // The Pythons you'd like to test against.  If not provided, defaults
+    // to the current version of Python used to run `asv`.
+    // "pythons": ["2.7", "3.6"],
+
+    // The list of conda channel names to be searched for benchmark
+    // dependency packages in the specified order
+    "conda_channels": ["conda-forge", "defaults"],
```
Comment (Cadair) [inline asv.conf.json:63]:
```suggestion
```

**Review by Cadair — COMMENTED** [review]
This looks like an excellent start, we should get it merged soon and we can always improve it once it's running.

I added a bunch of suggestions to clean out stuff we aren't using.

File: .github/workflows/asv-initial.yml [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+      - name: Add machine info for ASV
+        run : asv machine --machine GH-Actions --os ubuntu-18.04 --arch x64 --cpu "2-core unknown" --ram 7GB
+        
+      - name: Run benchmarks
+        run : asv run
+
+      - name: List results folder
+        run : |
+          cd asv_results/results
+          ls -la
```
Comment (nabobalis) [inline .github/workflows/asv-initial.yml:43]: Can we remove this?

File: .github/workflows/asv-initial.yml [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+          
+      - name: Push results
+        uses: JamesIves/github-pages-deploy-action@3.6.2
+        with:
+          BRANCH: main
+          FOLDER: asv_results/results
+          REPOSITORY_NAME: sunpy/sunpy-benchmarks
+          TARGET_FOLDER: results 
+          CLEAN: true
+          SSH: true # SSH must be set to true so the deploy action knows which protocol to deploy with.
```
Comment (nabobalis) [inline .github/workflows/asv-initial.yml:58]:
```suggestion
          SSH: true
```

File: .github/workflows/asv-initial.yml (hunk 1 above)
Comment (kathivashkiv) [reply .github/workflows/asv-initial.yml:11]: There is a note about this in the Actions doc:
```Note: The Ubuntu 20.04 virtual environment is currently provided as a preview only. The ubuntu-latest YAML workflow label still uses the Ubuntu 18.04 virtual environment.```
I use 18.04 (not `ubuntu-latest`) cause I give this version to ASV in the machine info

File: asv.conf.json (hunk 4 above)
Comment (kathivashkiv) [reply asv.conf.json:7]: No. ASV clones sunpy core repo into the directory named as this value in config. When I tried "sunpy" there was an error cause there is a "sunpy" folder, already. So there can be any name you want, but not "sunpy" as long as there is "sunpy" folder in core repo.

File: .github/workflows/asv-initial.yml (hunk 15 above)
Comment (kathivashkiv) [reply .github/workflows/asv-initial.yml:43]: Yes, we can, it was for debug

File: benchmarks/benchmarks_for_map.py (hunk 3 above)
Comment (kathivashkiv) [reply benchmarks/benchmarks_for_map.py:15]: Yes, it is recommended for timing benchmarks to take out of bounds all the preparations, to test only timing of function call

Comment (kathivashkiv) [reply benchmarks/benchmarks_for_map.py:15]: But it is possible that I misunderstood mechanism of using sample data in general

File: .github/workflows/asv-initial.yml (hunk 1 above)
Comment (Cadair) [reply .github/workflows/asv-initial.yml:11]: I would be surprised if it didn't work, don't think there is any drawbacks if it does work.

File: asv.conf.json (hunk 4 above)
Comment (Cadair) [reply asv.conf.json:7]: Ok, I would suggest "sunpy-benchmarks" or something then.

**Review by Cadair — COMMENTED** [review]
.

File: .github/workflows/asv-initial.yml (hunk 1 above)
Comment (kathivashkiv) [reply .github/workflows/asv-initial.yml:11]: Okay, I'll change version to 20.04

File: asv.conf.json (hunk 4 above)
Comment (kathivashkiv) [reply asv.conf.json:7]: Good

**Comment by kathivashkiv** [comment]
Done

**Comment by kathivashkiv** [comment]
Workflow run succeeded: https://github.com/kathivashkiv/sunpy/runs/1601495271

**Comment by kathivashkiv** [comment]
New result in the `sunpy-benchmarks` repo: https://github.com/sunpy/sunpy-benchmarks/commit/148e7039d29756f18375c4a716b098191f6759ee

File: .github/workflows/asv-initial.yml [hunk 17]
```diff
@@ -0,0 +1,58 @@
+name: asv-benchmarks
+
+on:
+  push:
+    branches: 
+      - asv-ci
```
Comment (nabobalis) [inline .github/workflows/asv-initial.yml:6]: This is still here as your fork has this branch?

File: benchmarks/benchmarks_for_parse_time.py [hunk 18]
```diff
@@ -0,0 +1,24 @@
+# Initial benchmarks
+# See "Writing benchmarks" in the asv docs for more information.
```
Comment (nabobalis) [inline benchmarks/benchmarks_for_parse_time.py:2]:
```suggestion
```

File: benchmarks/benchmarks_for_map.py [hunk 19]
```diff
@@ -0,0 +1,33 @@
+# Initial benchmarks
+# See "Writing benchmarks" in the asv docs for more information.
```
Comment (nabobalis) [inline benchmarks/benchmarks_for_map.py:2]:
```suggestion
```

**Comment by nabobalis** [comment]
You will need to run the style checks before we can merge this:

`tox -e codestyle`

File: .github/workflows/asv-initial.yml (hunk 17 above)
Comment (kathivashkiv) [reply .github/workflows/asv-initial.yml:6]: Yes. I can track if workflow runs for new commits in this pr are successful, or I broke smth

File: benchmarks/benchmarks_for_parse_time.py (hunk 18 above)
Comment (kathivashkiv) [reply benchmarks/benchmarks_for_parse_time.py:2]: I want to leave this "initial" comment cause this benchmarks are experimental, I don't know if they are good enough. So I want to mark them as "initial", only as examples to test ASV work. And when I'll write sets of benchmarks to cover sunpy modules, this two may be rewritten or deleted.

Comment (nabobalis) [reply benchmarks/benchmarks_for_parse_time.py:2]: I don't see the need to mark them. We know they are experimental.

**Comment by kathivashkiv** [comment]
> `tox -e codestyle`

Locally?

File: benchmarks/benchmarks_for_parse_time.py (hunk 18 above)
Comment (kathivashkiv) [reply benchmarks/benchmarks_for_parse_time.py:2]: Good

**Comment by nabobalis** [comment]
> > `tox -e codestyle`
> 
> Locally?

Yeah.

**Comment by kathivashkiv** [comment]
I used only pytest for sunpy before. I'll set up tox now

**Comment by kathivashkiv** [comment]
> > > `tox -e codestyle`

I have to run this in `sunpy-git` directory ?

**Comment by nabobalis** [comment]
> > > > `tox -e codestyle`
> 
> I have to run this in `sunpy-git` directory ?

Yeah, it should just work when you have tox installed.

**Comment by kathivashkiv** [comment]
There are some fails. I see this in my terminal:
```
pre-commit hook(s) made changes.
If you are seeing this message in CI, reproduce locally with: `pre-commit run --all-files`.
To run `pre-commit` as part of git workflow, use `pre-commit install`.
All changes made by hooks:
```

And in the end:
```
ERROR: InvocationError for command /home/kate/sunpy-git/.tox/codestyle/bin/pre-commit run --color always --all-files --show-diff-on-failure (exited with code 1)
___________________________________ summary ____________________________________
ERROR:   codestyle: commands failed
```

I'm not sure what to do now

**Comment by nabobalis** [comment]
Run it again, it should fix the issues it can and whats left over you need to fix yourself.

It reports the remaining issues in the output.

**Comment by kathivashkiv** [comment]
Done. New output is:
```
  codestyle: commands succeeded
  congratulations :)
```
And `git status` shows three files with changes not staged for commit. I have to add and commit them?

**Comment by nabobalis** [comment]
> Done. New output is:
> ```
>   codestyle: commands succeeded
>   congratulations :)
> ```
> And `git status` shows three files with changes not staged for commit. I have to add and commit them?

Exactly.

**Comment by kathivashkiv** [comment]
Done

File: benchmarks/benchmarks_for_map.py (hunk 3 above)
Comment (Cadair) [reply benchmarks/benchmarks_for_map.py:13]: So the reason I don't think this needs to be in setup is all `sunpy.data.sample.AIA_171_IMAGE` is is a string representing a file path. There is no time or memory cost to this, so why don't we just use the `AIA_171_IMAGE` constant in the benchmarks?

**Comment by Cadair** [comment]
The big change I would like to see here before we merge this is removing any parts of the config we aren't using?

**Comment by kathivashkiv** [comment]
> The big change I would like to see here before we merge this is removing any parts of the config we aren't using?

Okay, you convinced me. Can you delete all you want? It'll be easier than starting a conversation about it. If you do think, that some comments should be deleted -- delete them. Now I better understand ASV workflow, so now I think that I'll lose nothing, if you delete this stuff, now, in this config file.

File: asv.conf.json [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+
+    // The URL or local path of the source code repository for the
+    // project being benchmarked
+    "repo": "https://github.com/sunpy/sunpy",
+
+    // The Python project's subdirectory in your repo.  If missing or
+    // the empty string, the project is assumed to be located at the root
+    // of the repository.
+    // "repo_subdir": "",
+
```
Comment (Cadair) [inline asv.conf.json:20]:
```suggestion
```

**Review by nabobalis — APPROVED** [review]

**Pull request merged**